test/topology_random_failures/test_random_failures.py

```diff
@@ -0,0 +1,212 @@
+        coordinator = await get_coordinator_host(manager=manager)
+        await manager.api.enable_injection(node_ip=coordinator.ip_addr, injection=error_injection, one_shot=True)
+        coordinator_log = await manager.server_open_log(server_id=coordinator.server_id)
+        coordinator_log_mark = await coordinator_log.mark()
+        s_info = await manager.server_add(expected_server_state=ServerState.PROCESS_STARTED)
+        await coordinator_log.wait_for(
+            pattern="stop_after_updating_cdc_generation: wait for message for 5 minutes",
+            from_mark=coordinator_log_mark,
+        )
+        await manager.server_pause(server_id=s_info.server_id)
+        await manager.api.message_injection(node_ip=coordinator.ip_addr, injection=error_injection)
+    else:
+        s_info = await manager.server_add(
+            config={"error_injections_at_startup": [{"name": error_injection, "one_shot": True}]},
+            expected_server_state=ServerState.PROCESS_STARTED,
+        )
+
+    LOGGER.info("Wait till the bootstrapping node will be paused.")
+    await manager.wait_for_scylla_process_status(
+        server_id=s_info.server_id,
+        expected_statuses=[
+            psutil.STATUS_STOPPED,
+        ],
+    )
+
+    LOGGER.info("Run the cluster event main step.")
+    try:
+        cluster_event_start = time.perf_counter()
+        await anext(cluster_event_steps)
+        cluster_event_duration = time.perf_counter() - cluster_event_start
+        LOGGER.info("Cluster event `%s' took %.1fs", cluster_event.__name__, cluster_event_duration)
+    finally:
+        LOGGER.info("Unpause the server and wait till it wake up and become connectable using CQL.")
+        await manager.server_unpause(server_id=s_info.server_id)
+
+    server_log = await manager.server_open_log(server_id=s_info.server_id)
+
+    if cluster_event_duration + 1 >= WAIT_FOR_IP_TIMEOUT and error_injection in (  # give one more second for a tolerance
+        "stop_after_sending_join_node_request",
+        "stop_after_bootstrapping_initial_raft_configuration",
+    ):
+        LOGGER.info("Expecting the added node will hang and we'll have a message in the coordinator's log.  See #18638.")
+        coordinator = await get_coordinator_host(manager=manager)
+        coordinator_log = await manager.server_open_log(server_id=coordinator.server_id)
+        if matches := await coordinator_log.grep(r"The node may hang\. It's safe to shut it down manually now\."):
+            LOGGER.info("Found following message in the coordinator's log:\n\t%s", matches[-1][0])
+            await manager.server_stop(server_id=s_info.server_id)
+
+    if await manager.running_servers(s_info.server_id):
+        LOGGER.info("Wait until the new node initialization completes or fails.")
+        await server_log.wait_for("init - (Startup failed:|Scylla version .* initialization completed)", timeout=120)
+        time.sleep(1)
```

Comment:
Why is this sleep needed?

Reply:
Unfortunately, yes. The scylla process is not shutdown immediately after `Startup failed` message appears in the log. I saw flakiness in my runs because of that.

Reply:
So instead of sleeping for arbitrary time, why not `wait_for` the process to be dead?

Reply:
In general, it's because `server_log.wait_for()` doesn't return matched line:

https://github.com/scylladb/scylladb/blob/fe0e9618566e02a0128c3247ff3b11c5998b99ea/test/pylib/log_browsing.py#L42-L75

Reply:
so you can either modify `wait_for` to have similar interface to `grep` (which returns matches), OR after `wait_for` returns, call `grep` and it will return the match. Then if it's "failed", `wait_for` process to be dead.

`sleep(1)` is asking for flakiness

Reply:
OK, replaced it with additional log grep.